src/pudl/analysis/plant_parts_eia.py

```diff
@@ -1334,56 +1034,83 @@ def add_record_count_per_plant(self, part_df: pd.DataFrame) -> pd.DataFrame:
+        First the plant part list records are matched to generators. Then
+        the matched records are sorted by PLANT_PARTS_ORDERED and the
+        highest granularity record for each generator is marked as the true
+        granularity. The appropriate true granular part label and record id
+        is then merged on to get the plant part table with true granularity labels.
 
-        part_df = part_df.merge(
-            bool_df,
-            on=self.id_cols + IDX_TO_ADD + IDX_OWN_TO_ADD,
-            how="left",
-        ).rename(
+        Arguments:
+            ppl: (pd.DataFrame) The plant parts list
+        """
+        parts_to_gens = match_to_single_plant_part(
+            multi_gran_df=ppl,
+            ppl=ppl,
+            part_name="plant_gen",
+            cols_to_keep=["plant_part"],
+        )[["record_id_eia_og", "record_id_eia", "plant_part_og"]].rename(
             columns={
-                f"true_gran_{self.part_name}": "true_gran",
-                f"appro_part_label_{self.part_name}": "appro_part_label",
-                f"record_id_eia_{self.part_name}": "appro_record_id_eia",
+                "record_id_eia": "gen_id",
+                "record_id_eia_og": "record_id_eia",
+                "plant_part_og": "plant_part",
             }
         )
+        # concatenate the gen id's to get the combo of gens for each record
+        combos = (
+            parts_to_gens.groupby(["record_id_eia"])["gen_id"]
+            .apply(lambda x: ",".join(x))
+            .rename("gens_combo")
+        )
+        parts_to_gens = parts_to_gens.merge(
+            combos, how="left", left_on="record_id_eia", right_index=True
+        )
 
-        prop_true_len2 = len(part_df[part_df.true_gran]) / len(part_df)
-        logger.debug(f"proportion of trues: {prop_true_len2:.02}")
-        logger.debug(f"number of records post-merge: {len(part_df)}")
-        return part_df
+        # categorical columns allow sorting by PLANT_PARTS_ORDERED
+        parts_to_gens["plant_part"] = pd.Categorical(
+            parts_to_gens["plant_part"], PLANT_PARTS_ORDERED
+        )
+        parts_to_gens = parts_to_gens.sort_values("plant_part")
+        # get the true gran records by finding duplicate gen combos
+        dupes = parts_to_gens.duplicated(subset=["gens_combo"], keep="first")
+        parts_to_gens.loc[:, "true_gran"] = ~(dupes)
+        # drop duplicate record ids so there is one row for each record
+        parts_to_gens = parts_to_gens.drop_duplicates(subset=["record_id_eia"])
+        true_grans = (
+            parts_to_gens[parts_to_gens.true_gran][
+                ["record_id_eia", "plant_part", "gens_combo"]
+            ]
+            .rename(
+                columns={
+                    "record_id_eia": "appro_record_id_eia",
+                    "plant_part": "appro_part_label",
+                }
+            )
+            .astype({"appro_part_label": "object"})
```

Comment:
why the object type? is this bc it got assigned as a categorical earlier? would it be better to make it a string/a nullable string?

Reply:
Ya I made it an object because it was categorical earlier and I've run into all sorts of problems with categorical dtypes in the past (and I wanted it to pass the unit test I made which required the same dtype). I can change it to a string. Pandas dtypes are v confusing and I'm never sure if I should string or if I should object.

Reply:
100% fair! and yes the categorical dytpes are nice for some things but are a pain in the ass sometimes (like in groupbys omigosh `observed=True` should be the default). From my - albeit fuzzy - understanding the object dtype generally is less space efficient.

Reply:
That makes sense - `object` is maybe holding space for the column to be something other than a string.